pkgs/by-name/xa/xar/package.nix

```diff
@@ -0,0 +1,175 @@
+{
+  lib,
+  stdenv,
+  fetchurl,
+  applyPatches,
+
+  autoreconfHook,
+
+  # Required dependencies.
+  zlib,
+  libxml2,
+
+  # Optional dependencies.
+  withExt2 ? lib.meta.availableOn stdenv.hostPlatform e2fsprogs,
+  e2fsprogs,
+  withAcl ? lib.meta.availableOn stdenv.hostPlatform acl,
+  acl,
+  withBzip2 ? true,
+  bzip2,
+  withLzma ? true,
+  xz, # lzma
+
+  # Platform-specific dependencies.
+  musl-fts,
+  openssl,
+  darwin, # CommonCrypto
+
+  gitUpdater,
+
+  # for tests
+  xar,
+  testers,
+  python3,
+  libxslt, # xsltproc
+  runCommand,
+  makeWrapper,
+  buildPackages,
+}:
+let
+  pname = "xar";
+  version = "498";
+
+  pythonForTests = py: py.withPackages (p: [ p.xattr ]);
+
+  src = applyPatches {
+    name = "source";
+    src = fetchurl {
+      url = "https://opensource.apple.com/tarballs/xar/xar-${version}.tar.gz";
+      hash = "sha256-nO5PgLls9ZLMxUWk/dUeTaSlvTtHNJAWN9Z7BD7/PHU=";
+    };
+    patches = lib.filesystem.listFilesRecursive ./patches;
+  };
+in
+stdenv.mkDerivation {
+  inherit pname version src;
+
+  sourceRoot = "source/xar";
+
+  outputs = [
+    "out"
+    "lib"
+    "dev"
+  ];
+
+  strictDeps = true;
+
+  nativeBuildInputs = [ autoreconfHook ];
+
+  # For some reason libxml2 package headers are in subdirectory and thus aren’t
+  # picked up by stdenv’s C compiler wrapper (see ccWrapper_addCVars). This
+  # doesn’t really belong here and either should be part of libxml2 package or
+  # libxml2 in Nixpkgs can just fix their header paths.
+  env.NIX_CFLAGS_COMPILE = "-isystem ${libxml2.dev}/include/libxml2";
+
+  buildInputs =
+    [
+      zlib
+      libxml2
+    ]
+    # Darwin uses CommonCrypto, other platforms use openssl.
+    ++ lib.optional stdenv.hostPlatform.isDarwin darwin.CommonCrypto
+    ++ lib.optional (!stdenv.hostPlatform.isDarwin) openssl
+    ++ lib.optional withExt2 e2fsprogs
+    ++ lib.optional withAcl acl
+    ++ lib.optional withBzip2 bzip2
+    ++ lib.optional withLzma xz
+    ++ lib.optional stdenv.hostPlatform.isMusl musl-fts;
+
+  passthru.tests = {
+    version = testers.testVersion {
+      package = xar;
+      version = "1.8dev";
+    };
+
+    # Check that xar builds without optional dependencies.
+    minimal = xar.override {
+      withExt2 = false;
+      withAcl = false;
+      withBzip2 = false;
+      withLzma = false;
+    };
+
+    integrationTest =
+      let
+        python = pythonForTests python3.pythonOnBuildForHost;
+      in
+      runCommand "xar-tests-integration-test"
+        {
+          inherit src;
+          python = python.executable;
+          nativeBuildInputs = [
+            xar
+            python
+            libxslt
+          ];
+        }
+        ''
+          "$python" -- "$src"/xar/test/run-all.py
+          touch -- "$out"
+        '';
+
+    smokeTest =
+      runCommand "xar-tests-smoke-test"
+        {
+          inherit src;
+          depsBuildBuild = [
+            buildPackages.stdenv.cc
+            xar
```

Comment:
To be fair, this is how `testers.testVersion` behaves as well since it uses `nativeBuildInputs` and idiomatic usage is to pass the package fixed point from `callPackage` arguments (so it’s spliced).

This seems like a reasonable compromise between using an emulator for tests and not running the tests at all (or failing to do so because we can’t execute the program built for a different platform). Alternatively, we can define tests under `optionalAttrs (buildPlatform.canExecute hostPlatform)`, but I haven’t seen this done in any package.

We can’t run _all_ `xar` tests in Nix sandbox anyway (e.g. on Linux, ACL/EAs syscalls are filtered with seccomp), hence we expose `tests.manualTest` with `xar` built for the host platform.

Reply:
Fair enough. I’d personally lean towards avoiding a false sense of security from testing a version that isn’t actually to be deployed – especially since it results in an additional build of `xar` (albeit one that is likely cached) – but if it’s common for other packages to be doing this kind of thing and we don’t have an ideal solution I certainly won’t block on it.

It would be nice if we could just make the tests native to the host platform but condition them on the availability of a builder for that platform, which is more nuanced than `buildPlatform.canExecute`, but I guess that would be pretty impure.

Reply:
Since this conversation, I’ve seen a handful of packages that define tests under `optionalAttrs` so I went with your suggestion and wrapped the tests under `canExecute` for `finalAttrs.finalPackage`.